File: study/main1.py
# RPi.GPIO 모듈 (라즈베리파이 GPIO 핀 제어 라이브러리)
import RPi.GPIO as GPIO
import time

# GPIO 핀 번호 설정 (BCM 모드 사용)
GPIO.setmode(GPIO.BCM)

# GPIO 핀 16, 20, 21번을 출력 모드로 설정
GPIO.setup(16, GPIO.OUT)
GPIO.setup(20, GPIO.OUT)
GPIO.setup(21, GPIO.OUT)

try:
    while 1:  # 반복
        # GPIO 핀 16, 20, 21번에 HIGH 신호 출력 (LED 켜기)
        GPIO.output(16, GPIO.HIGH)
        GPIO.output(20, GPIO.HIGH)
        GPIO.output(21, GPIO.HIGH)
        time.sleep(1.0)  # 1초 대기
        
        # GPIO 핀 16, 20, 21번에 LOW 신호 출력 (LED 끄기)
        GPIO.output(16, GPIO.LOW)
        GPIO.output(20, GPIO.LOW)
        GPIO.output(21, GPIO.LOW)
        time.sleep(1.0)  

except KeyboardInterrupt:  # 키보드 인터럽트(Ctrl + C) 발생 시 실행
    pass  # 아무 작업도 하지 않고 넘어감

# GPIO 설정 초기화 (모든 핀을 안전한 상태로)
GPIO.cleanup()

File: Study_C/project_1/main_1.c
// libgpiod 설치 및 확인
// sudo apt install -y gpiod libgpiod-dev
// gpiodetect
// libgpiod 이외에도 sysfs와 WiringPi도 존재

#include <gpiod.h>  // libgpiod 헤더 파일 포함
#include <stdio.h>
#include <unistd.h>  // sleep 함수 사용을 위해

#define GPIO_CHIP "/dev/gpiochip0"  // GPIO 컨트롤러 디바이스 경로
#define LED1 16  // GPIO 핀 번호 (BCM 모드)
#define LED2 20  // GPIO 핀 번호 (BCM 모드)
#define LED3 21  // GPIO 핀 번호 (BCM 모드)

int main() {
    struct gpiod_chip *chip;
    struct gpiod_line *line1, *line2, *line3;
    int ret;

    // GPIO 컨트롤러 열기
    chip = gpiod_chip_open(GPIO_CHIP);
    if (!chip) {
        perror("Failed to open GPIO chip");
        return 1;
    }

    // GPIO 라인 가져오기
    line1 = gpiod_chip_get_line(chip, LED1);
    line2 = gpiod_chip_get_line(chip, LED2);
    line3 = gpiod_chip_get_line(chip, LED3);
    if (!line1 || !line2 || !line3) {
        perror("Failed to get GPIO line");
        gpiod_chip_close(chip);
        return 1;
    }

    // GPIO 라인을 출력으로 요청
    ret = gpiod_line_request_output(line1, "led_blink", 0);
    if (ret < 0) {
        perror("Failed to request line1 as output");
        gpiod_chip_close(chip);
        return 1;
    }
    ret = gpiod_line_request_output(line2, "led_blink", 0);
    if (ret < 0) {
        perror("Failed to request line2 as output");
        gpiod_chip_close(chip);
        return 1;
    }
    ret = gpiod_line_request_output(line3, "led_blink", 0);
    if (ret < 0) {
        perror("Failed to request line3 as output");
        gpiod_chip_close(chip);
        return 1;
    }

    printf("Press Ctrl+C to stop the program.\n");

    // LED 깜빡이기
    while (1) {
        printf("LEDs ON\n");
        gpiod_line_set_value(line1, 1);  // LED1 ON
        gpiod_line_set_value(line2, 1);  // LED2 ON
        gpiod_line_set_value(line3, 1);  // LED3 ON
        sleep(1);  

        printf("LEDs OFF\n");
        gpiod_line_set_value(line1, 0);  // LED1 OFF
        gpiod_line_set_value(line2, 0);  // LED2 OFF
        gpiod_line_set_value(line3, 0);  // LED3 OFF
        sleep(1);  
    }

    // 리소스 해제 (프로그램 종료 시)
    gpiod_line_release(line1);
    gpiod_line_release(line2);
    gpiod_line_release(line3);
    gpiod_chip_close(chip);

    return 0;
}
